compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/model/types/BUnionType.java

```diff
@@ -507,15 +511,17 @@ private void populateMemberSemTypes(BType memberType, LinkedHashSet<SemType> mem
             return;
         }
 
-        SemType s = SemTypeHelper.semType(memberType, ignoreTypeIds);
+        SemType s = memberType.semType();
         if (!Core.isNever(s)) {
             memberSemTypes.add(s);
         }
     }
 
-    public SemType semTypeIgnoringTypeIds() {
-        populateMemberSemTypes(true);
-        return computeResultantUnion(memberSemTypes);
+    /**
+     * When the type is mutated we need to reset resolved semType.
+     */
+    public void resetSemType() {
+        this.semType = null;
```

Comment:
Shouldn't we also clear the `memberSemTypes` as well?

Reply:
We repopulate the memberSemTypes here, whenever the semtype is `null`.

https://github.com/ballerina-platform/ballerina-lang/blob/efc16e5a61cbfa9ce2402527d1a34bba934ef2d6/compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/model/types/BUnionType.java#L485-L492